File: Algorithms/0086_Partition_List/Java/PartitionListSolution1.java

// Space: O(1)
// Time: O(n)

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode partition(ListNode head, int x) {
       if (head == null || head.next == null) return head;
        ListNode temp1 = new ListNode(0);
        ListNode temp2 = new ListNode(0);
        ListNode pointer1 = temp1;
        ListNode pointer2 = temp2;
        ListNode cur = head;

        while (cur != null) {
            if (cur.val < x) {
                pointer1.next = cur;
                pointer1 = pointer1.next;
            } else if (cur.val >= x) {
                pointer2.next = cur;
                pointer2 = pointer2.next;
            }
            cur = cur.next;
        }
        pointer2.next = null;
        pointer1.next = temp2.next;
        return temp1.next;
    }
}

File: Algorithms/0086_Partition_List/Python/Partition_List_Solution_1.py

# Space: O(1)
# Time: O(n)


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def partition(self, head, x):
        if head is None or head.next is None: return head
        temp1, temp2 = ListNode(0), ListNode(0)
        pointer1, pointer2 = temp1, temp2
        cur = head

        while cur:
            if cur.val < x:
                pointer1.next = cur
                pointer1 = pointer1.next
            elif cur.val >= x:
                pointer2.next = cur
                pointer2 = pointer2.next
            cur = cur.next

        pointer1.next = temp2.next
        pointer2.next = None

        return temp1.next





